### PR TITLE
GameCommands: Train Station Placement

### PR DESCRIPTION
**Pull request by duncanspumpkin**
~~Wip some bits still missing and a bit broken atm.~~

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
This should now be all working. Haven't fully tested it though.

**Comment by AaronVanGeffen** [comment]
Did you deliberately increase station lengths? I'm sure many players will be happy to see it, just thought I'd check.

Screenshot 2023-12-10 at 16 22 39

**Comment by AaronVanGeffen** [comment]
> The "station too spread out" error no longer occurs. Did you deliberately increase station lengths? 

The latest version (sadly) no longer exhibits this bug. Can we still get rid of this, deliberately this time, though? I think it'd be a nice feature for 23.12 actually.

**Comment by AaronVanGeffen** [comment]
> The "station too spread out" error no longer occurs. Did you deliberately increase station lengths? 

The latest version (sadly) no longer exhibits this behaviour. I don't mean to move the goal posts for this PR. Would you prefer to change things up in a follow-up PR? I'd be happy to file one to deliberately change this behaviour.

**Comment by duncanspumpkin** [comment]
> > The "station too spread out" error no longer occurs. Did you deliberately increase station lengths?
> 
> The latest version (sadly) no longer exhibits this behaviour. I don't mean to move the goal posts for this PR. Would you prefer to change things up in a follow-up PR? I'd be happy to file one to deliberately change this behaviour.

Yeah follow on PR for that.

File: src/OpenLoco/src/GameCommands/GameCommands.cpp [hunk 1]
```diff
@@ -211,6 +212,15 @@ namespace OpenLoco::GameCommands
             regs.ebx = backup.ebx;
             return 0;
         });
+
+        // Used by a gc_unk_51 of going via doCommand
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/GameCommands/GameCommands.cpp:216]:
```suggestion
        // Used by gc_unk_51 instead of going via doCommand
```

**Review by AaronVanGeffen — APPROVED** [review]
Building new train station tiles appears to be working properly, that is with vanilla limits.

**Comment by AaronVanGeffen** [comment]
Just to be sure: could you run the (AI) replay tests again as well?

**Comment by LeftofZen** [comment]
I found a small bug with this:

The station placement tool is incorrectly showing the station catchment grid when I mouse over a curve or other piece of track that cannot have a station on it. In this image, my cursor is on the curved piece of track (screenshot didn't capture cursor)

image

**Comment by duncanspumpkin** [comment]
> I found a small bug with this:
> 
> The station placement tool is incorrectly showing the station catchment grid when I mouse over a curve or other piece of track that cannot have a station on it.
> 
> In this image, my cursor is on the curved piece of track (screenshot didn't capture cursor)
> 
> image
> 
> Clicking to place the station actually works too and it charges you money. The whole curved track piece is then marked as station. image
> 
> image
> 
> This is actually a great feature (stations on curves and more pieces) but without graphics, sadly it's a bug :(

okay found the source of that and fixed it.

**Review by LeftofZen — APPROVED** [review]
Did a quick playtest, station placement now seems correct. Placed all station types with no issues.

**Comment by duncanspumpkin** [comment]
Getting a divergence. Not sure if its from master though.

**Comment by duncanspumpkin** [comment]
[MMTest100000_base.zip](https://github.com/OpenLoco/OpenLoco/files/13960831/MMTest100000_base.zip)
New reference save after running my MM ai test file for 100000 ticks.

**Review by AaronVanGeffen — APPROVED** [review]
Tested the most recent version. Station placement placement still working fine. AI companies also still spaghetti stations around, so business as usual there as well.

File: src/OpenLoco/src/GameCommands/Track/CreateTrainStation.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#include "GameCommands/GameCommands.h"
+
+namespace OpenLoco::GameCommands
+{
+    struct TrackStationPlacementArgs
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/GameCommands/Track/CreateTrainStation.h:5]: Should we rename this, for consistency? Can be done in another PR, I'm not too fussed.
```suggestion
    struct TrainStationPlacementArgs
```

Comment (duncanspumpkin) [reply src/OpenLoco/src/GameCommands/Track/CreateTrainStation.h:5]: Okay i've gone and renamed all the TrackStation things that are for train stations. There are still some trackstation things but that is the name that has been kept for things used by both road and track.

**Pull request merged**